commons/src/main/java/esa/commons/io/AbstractPathWatcher.java

```diff
@@ -0,0 +1,265 @@
+                pushEvent(new WatchEventContextImpl<>(event, file), delete);
+            } else if (kind == StandardWatchEventKinds.OVERFLOW) {
+                pushEvent(new WatchEventContextImpl<>(event, file), overflow);
+            }
+        }
+        wk.reset();
+    }
+
+    abstract void initDir(Path path);
+
+    abstract void register(Path path);
+
+    abstract File getFile(WatchEvent<?> event, WatchKey wk);
+
+    private void pushEvent(WatchEventContext<?> ctx, Consumer<WatchEventContext<?>> consumer) {
+        if (delay <= 0) {
+            consumer.accept(ctx);
+            return;
+        }
+
+        final String eventKey = ctx.file().getAbsolutePath() + ctx.event().kind().name();
+        if (eventSet.contains(eventKey)) {
+            return;
+        }
+        eventSet.add(eventKey);
+
+        //Use atomic operations to avoid unnecessary exceptions caused by adding tasks to
+        //delayScheduler when delayScheduler had shutdown after stop
+        atomicSchedulerOperation(() -> {
+            if (!stopped) {
+                delayScheduler.schedule(() -> {
+                    //Remove first and then call consumer.accept() to avoid the new event is removed
+                    //before it has been executed when the new event appears in the process of
+                    //consumer.accept().
+                    eventSet.remove(eventKey);
+                    consumer.accept(ctx);
+                }, delay, TimeUnit.MILLISECONDS);
+            }
+        });
+
+    }
+
+    private void atomicSchedulerOperation(Runnable runnable) {
+        Checks.checkNotNull(delayScheduler, "delayScheduler");
+        synchronized (delayScheduler) {
+            runnable.run();
+        }
+    }
+
+    private static Consumer<WatchEventContext<?>> doIfConsumerNotNull(Consumer<WatchEventContext<?>> consumer,
+                                                                      Runnable doIfConsumerNotNull) {
+        if (consumer == null) {
+            return null;
+        } else {
+            doIfConsumerNotNull.run();
+            return consumer;
+        }
+    }
+
+    private static ScheduledExecutorService delayScheduler(ScheduledExecutorService scheduler, Path path) {
+        if (scheduler == null) {
+            return defaultScheduler("FileWatcher-delayScheduler-" + path);
+        } else {
+            return scheduler;
+        }
+    }
+
+    /**
+     * The executor does not share threads to avoid the bad influence of FileWatchers among
+     * multiple components.
+     * <p>
+     * When the command execution ends, the thread resource will be recycled automatically.This means
+     * when the user stops FileWatcher, the thread resources will be automatically recycled without
+     * other processing.
+     *
+     * @param path path
+     * @return executor
+     */
+    private static Executor executor(Path path) {
+        return command -> new Thread(command, "FileWatcher-Executor-" + path).start();
```

Comment:
we expect that different users use a same watcher thread or just let end-user to choose it